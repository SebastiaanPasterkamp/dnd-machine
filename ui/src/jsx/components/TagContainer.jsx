import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import {
    TagsContainer,
    Tag,
    TagBadge,
} from './BaseTagContainer.jsx';
import LazyComponent from './LazyComponent.jsx';
import SingleSelect from './SingleSelect.jsx';

export class TagContainer extends LazyComponent
{
    onDelete(tag, index) {
        const { value, onDelete, setState } = this.props;

        setState(
            _.concat(
                _.slice(value, 0, index),
                _.slice(value, index + 1)
            )
        );

        if (onDelete) {
            onDelete(tag, index);
        }
    }

    onAdd(newValue) {
        const { value, onAdd, setState } = this.props;

        setState(
            _.concat(value, [newValue])
        );

        if (onAdd) {
            onAdd(newValue);
        }
    }

    renderSelect() {
        const {
            value, items = [], disabled, multiple, showSelect = true,
        } = this.props;

        if (disabled) return null;
        if (!showSelect) return null;
        if (!items.length) return null;

        const filtered = _.chain(items)
            .filter(item => (
                multiple
                || !_.includes(value,  _.get(item, 'code', item.name))
            ))
            .map(item => _.pickBy(
                {
                    code: _.get(item, 'code', item.name),
                    label: _.get(item, 'label', item.name),
                    description: item.description,
                },
                v => v !== undefined
            ))
            .value();

        if (!filtered.length) return null;

        return <SingleSelect
            emptyLabel="Add..."
            items={filtered}
            setState={item => this.onAdd(item)}
            />;
    }

    render() {
        const {
            value, items = [], disabled, multiple, showSelect = true,
           className,
        } = this.props;
        const counts = _.countBy(value);

        const tags = _.map(counts, (count, key) => {
            const { label, description } = (
                _.find(items, { code: key })
                || _.find(items, { name: key })
                || {}
            );

            return {
                key,
                count,
                label,
                description,
                disabled,
                onDelete: () => this.onDelete(
                    key,
                    _.indexOf(value, key)
                ),
            };
        });

        return <TagsContainer className={className}>
            {this.renderSelect()}

            {_.map(tags, tag => (
                <Tag {...tag}>
                    {multiple && tag.count > 1 &&
                        <TagBadge>
                            &times;&nbsp;{ tag.count }
                        </TagBadge>
                    }
                </Tag>
            ))}
        </TagsContainer>
    }
};

TagContainer.propTypes = {
    value: PropTypes.arrayOf(PropTypes.string).isRequired,
    setState: PropTypes.func.isRequired,
    className: PropTypes.string,
    showSelect: PropTypes.bool,
    multiple: PropTypes.bool,
    disabled: PropTypes.bool,
    items: PropTypes.arrayOf(
        PropTypes.object
    ),
    onAdd: PropTypes.func,
    onDelete: PropTypes.func,
};

export default TagContainer;
