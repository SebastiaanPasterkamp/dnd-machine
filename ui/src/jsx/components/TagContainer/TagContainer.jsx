import React from 'react';
import PropTypes from 'prop-types';
import {
    countBy,
    find,
    indexOf,
    keys,
    map,
    uniqBy,
} from 'lodash/fp';

import { memoize } from '../../utils';

import {
    BaseTagContainer,
    Tag,
    TagBadge,
} from '../BaseTagContainer';
import SingleSelect from '../SingleSelect';

import TagSelect from './components/TagSelect';

export class TagContainer extends React.Component
{
    constructor(props) {
        super(props);
        this.onAdd = this.onAdd.bind(this);
        this.memoize = memoize.bind(this);
    }

    onAdd(item) {
        const { value, onAdd, multiple, objects, setState } = this.props;
        const { type, id, name } = objects ? item : { id: item };
        const needle = objects ? { type, id } : item;
        const index = indexOf(needle, value);

        if (!multiple && index >= 0) {
            return;
        }

        if (onAdd) {
            onAdd(item);
        }

        if (!objects) {
            setState([
                ...value,
                item,
            ]);
            return;
        }

        if (index >= 0 && multiple) {
            setState([
                ...value.slice(0, index),
                {...value[index], count: (value[index].count || 1) + 1},
                ...value.slice(index + 1),
            ]);
            return;
        }

        if (multiple) {
            setState([
                ...value,
                { type, id, name, count: 1},
            ]);
            return;
        }

        setState([
            ...value,
            { type, id, name },
        ]);
    }

    onDelete(item) {
        const key = this.getKey(item);

        return this.memoize(key, () => {
            const { value, onDelete, multiple, objects, setState } = this.props;
            const { type, id, name } = objects ? item : { id: item };
            const needle = objects ? { type, id } : item;
            const index = indexOf(needle, value);

            if (onDelete) {
                onDelete(item, index);
            }

            if (multiple && objects && value[index].count > 0) {
                setState([
                    ...value.slice(0, index),
                    {...value[index], count: value[index].count - 1},
                    ...value.slice(index + 1),
                ]);
                return;
            }

            setState([
                ...value.slice(0, index),
                ...value.slice(index + 1),
            ]);
        });
    }

    getKey(item) {
        const { objects } = this.props;
        return objects ? `${item.type}.${item.id}` : item;
    }

    render() {
        const {
            value,
            items,
            disabled: propDisabled,
            multiple,
            objects,
            showSelect,
           className,
        } = this.props;
        const counts = countBy(item => this.getKey(item))(value);

        const tags = uniqBy('key', map(
            item => {
                const key = this.getKey(item);
                const {
                    id = item,
                    count = counts[key],
                    type, name, description, disabled,
                } = objects ? item : (
                    find({ id: item }, items)
                    || find({ name: item }, items)
                    || {}
                );

                return {
                    key,
                    type,
                    id,
                    count,
                    name,
                    description,
                    disabled: propDisabled || disabled,
                    onDelete: this.onDelete(item),
                };
            }
        )(value));

        return (
            <BaseTagContainer className={className}>
                {!propDisabled && showSelect ? (
                    <TagSelect
                        onSelect={this.onAdd}
                        items={items}
                        current={value}
                        objects={objects}
                        multiple={multiple}
                    />
                ) : null}

                {map(
                    tag => (
                        <Tag {...tag}>
                            {multiple ? (
                                <TagBadge>
                                    &times;&nbsp;{ tag.count }
                                </TagBadge>
                            ) : null}
                        </Tag>
                    )
                )(tags)}
            </BaseTagContainer>
        );
    }
};

TagContainer.propTypes = {
    value: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
            PropTypes.shape({
                type: PropTypes.string,
                id: PropTypes.oneOfType([
                    PropTypes.string,
                    PropTypes.number,
                    PropTypes.bool,
                ]).isRequired,
                name: PropTypes.string.isRequired,
            }),
        ])
    ).isRequired,
    setState: PropTypes.func,
    className: PropTypes.string,
    showSelect: PropTypes.bool,
    multiple: PropTypes.bool,
    objects: PropTypes.bool,
    disabled: PropTypes.bool,
    items: PropTypes.arrayOf(
        PropTypes.object
    ),
    onAdd: PropTypes.func,
    onDelete: PropTypes.func,
};

TagContainer.defaultProps = {
    value: [],
    setState: null,
    className: null,
    showSelect: true,
    multiple: false,
    objects: false,
    disabled: false,
    items: [],
    onAdd: null,
    onDelete: null,
};

export default TagContainer;
