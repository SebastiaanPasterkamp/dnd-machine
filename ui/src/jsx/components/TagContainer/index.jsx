export { default, TagContainer } from './TagContainer';
