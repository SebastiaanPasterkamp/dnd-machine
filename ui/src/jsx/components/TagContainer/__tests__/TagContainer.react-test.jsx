import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import TagContainer from '..';

const items = [
    { id: "a", name: "Aaa" },
    { id: "b", name: "Bbb", description: "Bee" },
];

describe('Component: TagContainer', () => {
    it('should render with minimal props', () => {
        const setState = jest.fn();
        const tree = renderer.create(
            <TagContainer
                setState={setState}
                value={[]}
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render with all props', () => {
        const setState = jest.fn();
        const tree = renderer.create(
            <TagContainer
                setState={setState}
                items={items}
                value={['a', 'b', 'a']}
                className="info"
                disabled={true}
                multiple={true}
                showSelect={false}
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should exclude selected items from select in single mode', () => {
        const setState = jest.fn();
        const tree = renderer.create(
            <TagContainer
                setState={setState}
                items={items}
                value={['a']}
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should allow all items in select in multiple mode', () => {
        const setState = jest.fn();
        const tree = renderer.create(
            <TagContainer
                setState={setState}
                items={items}
                value={['a']}
                multiple={true}
                />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should emit changes when adding', () => {
        const setState = jest.fn();
        const onAdd = jest.fn();
        const wrapper = mount(
            <TagContainer
                setState={setState}
                onAdd={onAdd}
                items={items}
                value={['a']}
                multiple={true}
                />
        );

        wrapper
            .find('button.nice-btn')
            .simulate('click');
        wrapper
            .find('li[data-value="b"] a')
            .simulate('click');

        expect(setState)
            .toBeCalledWith(['a', 'b']);
        expect(onAdd)
            .toBeCalledWith('b');
    });

    it('should emit changes when deleting', () => {
        const setState = jest.fn();
        const onDelete = jest.fn();
        const wrapper = mount(
            <TagContainer
                setState={setState}
                onDelete={onDelete}
                items={items}
                value={['a', 'b']}
                />
        );

        wrapper
            .find('button.nice-tag-btn')
            .at(0)
            .simulate('click');

        expect(setState)
            .toBeCalledWith(['b']);
        expect(onDelete)
            .toBeCalledWith('a', 0);
    });
});
