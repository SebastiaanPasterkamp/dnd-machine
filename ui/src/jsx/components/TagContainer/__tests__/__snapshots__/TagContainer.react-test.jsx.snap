// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TagContainer should allow all items in select in multiple mode 1`] = `
<div
  className="base-tag-container nice-tags-container"
>
  <div
    className="nice-dropdown nice-form-control"
  >
    <button
      className="nice-btn"
      onClick={[Function]}
    >
      Add...
      <i
        className="icon fa-angle-down"
      />
    </button>
    <ul
      className="dropdown-menu"
      onClick={[Function]}
    >
      <li
        className={null}
        data-value="a"
        onClick={[Function]}
      >
        <a>
          Aaa
        </a>
      </li>
      <li
        className={null}
        data-value="b"
        onClick={[Function]}
      >
        <a>
          Bbb
        </a>
      </li>
    </ul>
  </div>
  <div
    className="nice-tag"
  >
    <span
      className="nice-tag-label"
    >
      Aaa
    </span>
    <span
      className="nice-tag-badge"
    >
      × 
      1
    </span>
    <button
      className="nice-tag-btn"
      onClick={[Function]}
    >
      <i
        className="icon fa-trash-o"
      />
    </button>
  </div>
</div>
`;

exports[`Component: TagContainer should exclude selected items from select in single mode 1`] = `
<div
  className="base-tag-container nice-tags-container"
>
  <div
    className="nice-dropdown nice-form-control"
  >
    <button
      className="nice-btn"
      onClick={[Function]}
    >
      Add...
      <i
        className="icon fa-angle-down"
      />
    </button>
    <ul
      className="dropdown-menu"
      onClick={[Function]}
    >
      <li
        className={null}
        data-value="b"
        onClick={[Function]}
      >
        <a>
          Bbb
        </a>
      </li>
    </ul>
  </div>
  <div
    className="nice-tag"
  >
    <span
      className="nice-tag-label"
    >
      Aaa
    </span>
    <button
      className="nice-tag-btn"
      onClick={[Function]}
    >
      <i
        className="icon fa-trash-o"
      />
    </button>
  </div>
</div>
`;

exports[`Component: TagContainer should render with all props 1`] = `
<div
  className="base-tag-container nice-tags-container info"
>
  <div
    className="nice-tag muted"
  >
    <span
      className="nice-tag-label"
    >
      Aaa
    </span>
    <span
      className="nice-tag-badge"
    >
      × 
      2
    </span>
  </div>
  <div
    className="nice-tag muted"
  >
    <span
      className="nice-tag-label"
    >
      <div
        className="tool-tip inline"
      >
        Bbb
        <span
          className="tool-tip--content"
        >
          <p>
            Bee
          </p>
        </span>
      </div>
    </span>
    <span
      className="nice-tag-badge"
    >
      × 
      1
    </span>
  </div>
</div>
`;

exports[`Component: TagContainer should render with minimal props 1`] = `
<div
  className="base-tag-container nice-tags-container"
/>
`;
